src/sender_queue/mod.rs

```diff
@@ -0,0 +1,366 @@
+    ) -> bool {
+        let pass =
+            |&them: &<D::Message as Epoched>::Epoch| self.is_accepting_epoch(&msg.message, them);
+        match &msg.target {
+            Target::All => peer_epochs.values().all(pass),
+            Target::Node(id) => peer_epochs.get(&id).map_or(false, pass),
+        }
+    }
+
+    /// A _spanning epoch_ of an epoch `e` is an epoch `e0` such that
+    ///
+    /// - `e` and `e0` are incomparable by the partial ordering on epochs and
+    ///
+    /// - the duration of `e0` is at least that of `e`.
+    ///
+    /// Returned is a list of spanning epochs for the given epoch.
+    ///
+    /// For example, any `DynamicHoneyBadger` epoch `Epoch((x, Some(y)))` has a unique spanning
+    /// epoch `Epoch((x, None))`. In turn, no epoch `Epoch((x, None))` has a spanning epoch.
+    fn spanning_epochs(
+        epoch: <D::Message as Epoched>::Epoch,
+    ) -> Vec<<D::Message as Epoched>::Epoch>;
+}
+
+pub type OutgoingQueue<D> = BTreeMap<
+    (
+        <D as DistAlgorithm>::NodeId,
+        <<D as DistAlgorithm>::Message as Epoched>::Epoch,
+    ),
+    Vec<<D as DistAlgorithm>::Message>,
+>;
+
+/// An instance of `DistAlgorithm` wrapped with a queue of outgoing messages, that is, a sender
+/// queue. This wrapping ensures that the messages sent to remote instances lead to progress of the
+/// entire consensus network. In particular, messages to lagging remote nodes are queued and sent
+/// only when those nodes' epochs match the queued messages' epochs. Thus all nodes can handle
+/// incoming messages without queueing them and can ignore messages whose epochs are not currently
+/// acccepted.
+#[derive(Debug)]
+pub struct SenderQueue<D>
+where
+    D: DistAlgorithm + SenderQueueFunc<D>,
+    D::Message: Clone + Epoched + Serialize + DeserializeOwned,
+    D::NodeId: NodeIdT + Rand,
+{
+    /// The managed `DistAlgorithm` instance.
+    algo: D,
+    /// Our node ID.
+    our_id: D::NodeId,
+    /// The set of all remote nodes on the network including validator as well as non-validator
+    /// (observer) nodes.
+    peer_ids: BTreeSet<D::NodeId>,
+    /// Current epoch.
+    epoch: <D::Message as Epoched>::Epoch,
+    /// Messages that couldn't be handled yet by remote nodes.
+    outgoing_queue: OutgoingQueue<D>,
+    /// Known current epochs of remote nodes.
+    peer_epochs: BTreeMap<D::NodeId, <D::Message as Epoched>::Epoch>,
+}
+
+pub type Step<D> = ::Step<SenderQueue<D>>;
+
+pub type Result<T, D> = ::std::result::Result<T, <D as DistAlgorithm>::Error>;
+
+impl<D> DistAlgorithm for SenderQueue<D>
+where
+    D: DistAlgorithm + Debug + Send + Sync + SenderQueueFunc<D>,
+    D::Message: Clone + Epoched + Serialize + DeserializeOwned,
+    D::NodeId: NodeIdT + Rand,
+{
+    type NodeId = D::NodeId;
+    type Input = D::Input;
+    type Output = D::Output;
+    type Message = Message<D::Message>;
+    type Error = D::Error;
+
+    fn handle_input(&mut self, input: Self::Input) -> Result<Step<D>, D> {
```

Comment:
Even for the `SenderQueue`, it would be nice to be usable without importing `DistAlgorithm`. Let's move the implementations of `handle_input` and `handle_message` to the `impl<D> SenderQueue<D>` below, and just call those from here.